pkg/apiserver/webhooks.go

```diff
@@ -226,14 +237,76 @@ func createValidatingWebhook(ctx context.Context, webhookNs string, caCert []byt
 	return nil
 }
 
-func runWebhookServer(certStorageDir string) error {
+// load the certificate & keep note of time loaded for reload on new cert details
+func loadCertificate(certPath, keyPath string) (tls.Certificate, error) {
+	certInfo, err := os.Stat(certPath)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+	if certInfo.ModTime().After(certModTime) {
+		newCert, err := tls.LoadX509KeyPair(certPath, keyPath)
+		if err != nil {
+			return tls.Certificate{}, err
+		}
+		cert = newCert
+		certModTime = certInfo.ModTime()
+	}
+	return cert, nil
+}
+
+// watch for changes on the mount path of the secret as volume
+func watchCertificates(directory, certFile, keyFile string) {
+	//set up a watcher for the cert
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.Fatalf("failed to start certificate watcher: %v", err)
+	}
+	defer watcher.Close()
+	// start the watcher with the dir to watch
+	err = watcher.Add(directory)
+	if err != nil {
+		klog.Errorf("Error in running watcher: %v", err)
+	}
+	// if the watcher notices any changes on the mount dir of the secret such as creations or writes to the files in this dir
+	// attempt to load tls cert using the new cert and key files provided and log output
+	done := make(chan bool)
+	go func() {
```

Comment:
This goroutine will prevent the whole application from gracefully shutting down. There should be a ctx.Done() channel listen in this one as well. Also the process should get the main context propagated down to it.